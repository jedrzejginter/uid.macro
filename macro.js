const { createMacro } = require('babel-plugin-macros');
const randomstring = require('randomstring');
const uuid = require('uuid');

// caching fixes issue with server-side rendering
// when file is being compiled two times
// without cache we get different uuids in server and client bundles
// which can cause hydration issues
const cache = {};

function genAndCacheValue(path, util, generator) {
  const key = `${path.hub.file.opts.filename}:${util}:${path.node.loc.start.line}.${path.node.loc.start.column}`;

  const val = cache[key] || generator();
  cache[key] = val;

  return val;
}

function uidMacro({ references, babel }) {
  const { str: strReferences = [], uuid: uuidReferences = [] } = references;
  const { types: t } = babel;

  strReferences.forEach((path) => {
    if (t.isCallExpression(path.parent)) {
      // handle optional length argument
      const [arg0] = path.parent.arguments;
      const len = t.isNumericLiteral(arg0) ? arg0.value : 24;

      const val = genAndCacheValue(path, 'str', () => randomstring(len));

      path.parentPath.replaceWith(t.stringLiteral(val));
    }
  });

  uuidReferences.forEach((path) => {
    if (t.isCallExpression(path.parent)) {
      const val = genAndCacheValue(path, 'uuid', uuid.v4);
      path.parentPath.replaceWith(t.stringLiteral(val));
    }
  });
}

module.exports = createMacro(uidMacro);
