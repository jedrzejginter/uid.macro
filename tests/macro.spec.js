const pluginTester = require('babel-plugin-tester').default;
const plugin = require('babel-plugin-macros');

// We have to mock 'uuid' module, because it would
// break snapshot every time we run tests, due to new
// uuid being generated.
jest.mock('uuid', () => {
  let i = 0;
  const uuids = [
    '00000000-0000-0000-0000-000000000001',
    '00000000-0000-0000-0000-000000000002',
  ];

  return {
    v4: () => {
      const value = uuids[i];
      i = (i + 1) % uuids.length;

      return value;
    }
  }
})

pluginTester({
  plugin,
  pluginName: 'uuid.macro',
  snapshot: true,
  babelOptions: {
    filename: __filename,
    plugins: ['@babel/plugin-syntax-jsx'],
  },
  tests: [
    `
      import { uuid } from '../macro';

      const x = uuid();

      function Component() {
        return <div id={uuid()} />;
      }
    `,
  ],
})
