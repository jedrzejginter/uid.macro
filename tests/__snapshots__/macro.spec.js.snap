// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uid.macro 1. uid.macro: 1. uid.macro 1`] = `

import { uuid } from '../macro';

const x = uuid();

function Component() {
  return <div id={uuid()} />;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const x = "00000000-0000-0000-0000-000000000001";

function Component() {
  return <div id={"00000000-0000-0000-0000-000000000002"} />;
}


`;

exports[`uid.macro 2. uid.macro: 2. uid.macro 1`] = `

import { str } from '../macro';

const x = str();

function Component() {
  return <div id={str(12)} />;
}

      ↓ ↓ ↓ ↓ ↓ ↓

const x = "firstrandomstring0123456";

function Component() {
  return <div id={"secondrandom"} />;
}


`;
